scheduler/scheduler.go

```diff
@@ -132,70 +133,70 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
-		for i := 0; i < numHostsToSpawn; i++ {
-			if ctx.Err() != nil {
-				return nil, errors.New("scheduling run canceled.")
+		// if distro can have containers, check if there are enough parents to hold
+		// new containers
+		if d.MaxContainers > 0 {
+			currentParents, err := host.FindAllRunningParents()
+			if err != nil {
+				return nil, errors.Wrap(err, "could not find running parents")
 			}
-
-			d, err := distro.FindOne(distro.ById(distroId))
+			existingContainers, err := host.FindAllRunningContainers()
 			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "failed to find distro",
-				}))
-				continue
+				return nil, errors.Wrap(err, "could not find running containers")
 			}
 
-			allDistroHosts, err := host.Find(host.ByDistroId(distroId))
-			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "problem getting hosts for distro",
-				}))
+			numNewParents := numNewParentsNeeded(len(currentParents), numHostsToSpawn,
+				len(existingContainers), d)
 
-				continue
-			}
+			// only want to spawn amount of parents allowed based on pool size
+			numNewParentsToSpawn := parentCapacity(d, numNewParents, len(currentParents), len(existingContainers), numHostsToSpawn)
+			// if there are parents to spawn, do not spawn containers in the same run
+			if numNewParentsToSpawn > 0 {
+				parentIntentHosts, err := insertParents(d, numNewParentsToSpawn)
+				if err != nil {
+					return nil, err
+				}
+				hostsSpawnedPerDistro[distroId] = append(hostsSpawnedPerDistro[distroId], parentIntentHosts...)
 
-			if len(allDistroHosts) >= d.PoolSize {
 				grip.Info(message.Fields{
-					"distro":    distroId,
-					"runner":    RunnerName,
-					"pool_size": d.PoolSize,
-					"message":   "max hosts running",
+					"runner":         RunnerName,
+					"distro":         distroId,
+					"num_parents":    numNewParentsToSpawn,
+					"num_containers": numHostsToSpawn,
+					"operation":      "spawning new parents",
+					"span":           time.Since(distroStartTime).String(),
+					"duration":       time.Since(distroStartTime),
 				})
-
-				continue
+				return hostsSpawnedPerDistro, nil
 			}
 
-			hostOptions := cloud.HostOptions{
-				UserName: evergreen.User,
-				UserHost: false,
-			}
-
-			intentHost := cloud.NewIntent(d, d.GenerateName(), d.Provider, hostOptions)
-			if err := intentHost.Insert(); err != nil {
-				err = errors.Wrapf(err, "Could not insert intent host '%s'", intentHost.Id)
-
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":   distroId,
-					"runner":   RunnerName,
-					"host":     intentHost.Id,
-					"provider": d.Provider,
-				}))
+			// only want to spawn amount of hosts/containers abased on
+			numHostsToSpawn = containerCapacity(d, len(currentParents), len(existingContainers), numHostsToSpawn)
```

Comment:
Can you call this numContainersToStart to avoid confusion?

Reply:
So I use the same name so that I can use it in the for loop underneath (since making regular hosts and containers can be done the same way). If I change the name, I can only think of duplicating the code for the for loop and use `numContainersToStart`. Is there a more elegant way to do this if I do change the name?